core/src/main/java/com/linecorp/armeria/common/multipart/BodyPartHeadersBuilder.java

```diff
@@ -0,0 +1,117 @@
+
+    private final HttpHeadersBuilder headers = HttpHeaders.builder();
+
+    @Nullable
+    private ContentDisposition contentDisposition;
+
+    private boolean isContentTypeSet;
+
+    BodyPartHeadersBuilder() {}
+
+    /**
+     * Adds a new header with the specified {@code name} and {@code value}.
+     */
+    public BodyPartHeadersBuilder add(String name, String value) {
+        return header(name, value, true);
+    }
+
+    /**
+     * Sets a header with the specified name and value. Any existing headers with the same name are
+     * overwritten.
+     */
+    public BodyPartHeadersBuilder set(CharSequence name, String value) {
+        return header(name, value, false);
+    }
+
+    private BodyPartHeadersBuilder header(CharSequence name, String value, boolean add) {
+        requireNonNull(name, "name");
+        requireNonNull(value, "value");
+        if (HttpHeaderNames.CONTENT_TYPE.contentEqualsIgnoreCase(name)) {
+            isContentTypeSet = true;
+        }
+        if (HttpHeaderNames.CONTENT_DISPOSITION.contentEqualsIgnoreCase(name)) {
+            contentDisposition = ContentDisposition.parse(value);
+        }
+
+        if (add) {
+            headers.add(name, value);
+        } else {
+            headers.set(name, value);
+        }
+        return this;
+    }
+
+    /**
+     * Adds the specified {@code Content-Type} header.
+     */
+    public BodyPartHeadersBuilder contentType(MediaType contentType) {
+        requireNonNull(contentType, "contentType");
+        isContentTypeSet = true;
+        return add(HttpHeaderNames.CONTENT_TYPE.toString(), contentType.toString());
+    }
+
+    /**
+     * Adds the specified {@code Content-Disposition} header.
+     */
+    public BodyPartHeadersBuilder contentDisposition(ContentDisposition contentDisposition) {
+        requireNonNull(contentDisposition, "contentDisposition");
+        this.contentDisposition = contentDisposition;
+        return add(HttpHeaderNames.CONTENT_DISPOSITION.toString(), contentDisposition.toString());
+    }
+
+    /**
+     * Returns a newly created {@link BodyPartHeaders} with the entries in this builder.
+     */
+    public BodyPartHeaders build() {
+        if (!isContentTypeSet) {
+            contentType(defaultContentType());
+        }
+        return new DefaultBodyPartHeaders(headers.build());
+    }
+
+    /**
+     * Returns the default {@code Content-Type} header value:
+     * {@link MediaType#OCTET_STREAM} if the
+     * {@code Content-Disposition} header is present with a non empty value,
+     * otherwise {@link MediaType#PLAIN_TEXT}.
+     *
+     * @see <a href="https://tools.ietf.org/html/rfc7578#section-4.4">RFC-7578</a>
+     */
+    private MediaType defaultContentType() {
+        if (contentDisposition != null && contentDisposition.filename() != null) {
+            return MediaType.OCTET_STREAM;
+        } else {
+            return MediaType.PLAIN_TEXT;
+        }
+    }
+}
```

Comment:
Once we replace `BodyPartHeaders` with `HttpHeaders`, we could move this logic somewhere else, like `MultiPartEncoder`.